Performance/perf_runner.sh

```diff
@@ -0,0 +1,212 @@
+    float, double, string,
+    ...and repeated variants of the above.
+
+Options:
+    -p <true|false>: Adds a packed option to each field.
+    -s[23]:          Generate proto2 or proto3 syntax. proto3 is
+                     the default.
+EOF
+  exit 1
+}
+
+function mkdir_if_necessary() {
+  mkdir "$@" 2>/dev/null || true
+}
+
+# ---------------------------------------------------------------
+# Functions for generating the .proto file.
+
+function print_proto_field() {
+  num="$1"
+  if [[ "$proto_syntax" == "2" ]] && [[ "$field_type" != repeated* ]]; then
+    type="optional $2"
+  else
+    type="$2"
+  fi
+
+  if [[ -n "$packed" ]]; then
+    echo "  $type field$num = $num [packed=$packed];"
+  else
+    echo "  $type field$num = $num;"
+  fi
+}
+
+function generate_test_proto() {
+  cat >"$gen_message_path" <<EOF
+syntax = "proto$proto_syntax";
+
+message PerfMessage {
+EOF
+
+  for field_number in $(seq 1 "$field_count"); do
+    print_proto_field "$field_number" "$field_type" >>"$gen_message_path"
+  done
+
+  cat >>"$gen_message_path" <<EOF
+}
+EOF
+}
+
+# ---------------------------------------------------------------
+# Functions for generating the Swift harness.
+
+function print_swift_set_field() {
+  num=$1
+  type=$2
+
+  case "$type" in
+    repeated\ string)
+      echo "        for _ in 0..<repeatedCount {"
+      echo "          msg.field$num.append(\"$((200+num))\")"
+      echo "        }"
+      ;;
+    repeated\ *)
+      echo "        for _ in 0..<repeatedCount {"
+      echo "          msg.field$num.append($((200+num)))"
+      echo "        }"
+      ;;
+    string)
+      echo "        msg.field$num = \"$((200+num))\""
+      ;;
+    *)
+      echo "        msg.field$num = $((200+num))"
+      ;;
+  esac
+}
+
+function generate_perf_harness() {
+  cat >"$gen_harness_path" <<EOF
+extension Harness {
+  func run() {
+    measure {
+      // Loop enough times to get meaningfully large measurements.
+      for _ in 0..<200 {
+        var msg = PerfMessage()
+EOF
+
+  for field_number in $(seq 1 "$field_count"); do
+    print_swift_set_field "$field_number" "$field_type" >>"$gen_harness_path"
+  done
+
+  cat >>"$gen_harness_path" <<EOF
+        let data = try msg.serializeProtobuf()
+        msg = try PerfMessage(protobuf: data)
+      }
+    }
+  }
+}
+EOF
+}
+
+# ---------------------------------------------------------------
+# Script main logic.
+
+packed=""
+proto_syntax="3"
+
+while getopts "p:s:" arg; do
+  case "${arg}" in
+    p)
+      packed="$OPTARG"
+      ;;
+    s)
+      proto_syntax="$OPTARG"
+      ;;
+  esac
+done
+shift "$((OPTIND-1))"
+
+if [[ "$proto_syntax" != "2" ]] && [[ "$proto_syntax" != "3" ]]; then
+  usage
+fi
+
+if [[ "$#" -ne 2 ]]; then
+  usage
+fi
+
+readonly field_count=$1
+readonly field_type=$2
+readonly script_dir="$(dirname $0)"
+
+# Make sure the runtime and plug-in are up to date first.
+pushd "$script_dir/.." >/dev/null
+swift build -c release
+popd >/dev/null
```

Comment:
Good one. I don't do as much shell scripting as I probably should, so tips like this are always appreciated.